File: tools/pack_web.py
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
OUT = ROOT / "firmware" / "HomeGate" / "webpage.h"
DELIM = ")===="


def dump(var: str, text: str) -> str:
    if DELIM in text:
        raise SystemExit(f"Refusing to pack {var}: delimiter found in file")
    return f'const char {var}[] PROGMEM = R"====({text})====";\n\n'


files = {
    "APP_INDEX": ROOT / "index.html",
    "APP_CSS": ROOT / "css" / "styles.css",
    "APP_JS": ROOT / "js" / "app.js",
    "APP_MANIFEST": ROOT / "manifest.json",
}

parts = [
    "#pragma once\n",
    "#include <pgmspace.h>\n\n",
]
for var, path in files.items():
    parts.append(dump(var, path.read_text(encoding="utf-8")))

OUT.parent.mkdir(parents=True, exist_ok=True)
OUT.write_text("".join(parts), encoding="utf-8")
print(f"Wrote {OUT}")

File: tools/pack_web.js
const fs = require("fs");
const path = require("path");

const root = path.resolve(__dirname, "..");
const out = path.join(root, "firmware", "HomeGate", "webpage.h");
const delim = ")====";

function dump(name, file) {
  const text = fs.readFileSync(file, "utf8");
  if (text.includes(delim)) {
    throw new Error(`Delimiter found in ${file}`);
  }
  return `const char ${name}[] PROGMEM = R"====(${text})====";\n\n`;
}

const parts = [
  "#pragma once\n",
  "#include <pgmspace.h>\n\n",
  dump("APP_INDEX", path.join(root, "index.html")),
  dump("APP_CSS", path.join(root, "css", "styles.css")),
  dump("APP_JS", path.join(root, "js", "app.js")),
  dump("APP_MANIFEST", path.join(root, "manifest.json")),
];

fs.mkdirSync(path.dirname(out), { recursive: true });
fs.writeFileSync(out, parts.join(""), "utf8");
console.log(`Wrote ${out}`);
